### PR TITLE
[cxx-interop] Do not treat `std::pair<UnsafeType, T>` as safe

### PR DESCRIPTION
**Pull request by egorzhdan**
rdar://109529750

### CODE REVIEW

**Comment by egorzhdan** [comment]
@swift-ci please smoke test

**Comment by egorzhdan** [comment]
This is required for https://github.com/apple/swift/pull/65087

File: lib/ClangImporter/ClangImporter.cpp [hunk 1]
```diff
@@ -6565,6 +6565,17 @@ static bool hasDestroyTypeOperations(const clang::CXXRecordDecl *decl) {
   return false;
 }
 
+static bool hasCustomCopyOrMoveConstructor(const clang::CXXRecordDecl *decl) {
+  // std::pair and std::tuple might have copy and move constructors, but that
+  // doesn't mean they are safe to use from Swift, e.g. std::pair<UnsafeType, T>
+  if (decl->isInStdNamespace() &&
+      (decl->getName() == "pair" || decl->getName() == "tuple")) {
```
Comment (zoecarver) [inline lib/ClangImporter/ClangImporter.cpp:6572]: Could we be more specific here?

Comment (egorzhdan) [reply lib/ClangImporter/ClangImporter.cpp:6581]: I don't think we should check fully qualified names if that's what you mean. For instance in libc++ pair is actually `std::__1::pair`, other stdlibs would have different fully qualified names.
Do you ideas on how to make this condition more reliable?

**Comment by egorzhdan** [comment]
@swift-ci please smoke test

**Review by zoecarver — APPROVED** [review]
Makes sense. Thank you!

**Pull request merged**